sql-plugin/src/main/scala/com/nvidia/spark/rapids/RmmRapidsRetryIterator.scala

```diff
@@ -0,0 +1,285 @@
+/*
+ * Copyright (c) 2023, NVIDIA CORPORATION.
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *     http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+package com.nvidia.spark.rapids
+
+import scala.collection.mutable
+
+import com.nvidia.spark.rapids.RapidsPluginImplicits._
+import com.nvidia.spark.rapids.jni.{RetryOOM, RmmSpark, SplitAndRetryOOM}
+
+import org.apache.spark.sql.vectorized.ColumnarBatch
+
+object RmmRapidsRetryIterator extends Arm {
+  def splitInHalfByRows(
+      spillable: SpillableColumnarBatch): Seq[SpillableColumnarBatch] = {
+    val toSplitRows = spillable.numRows
+    if (toSplitRows <= 1) {
```

Comment:
nit: I know that this is the actual limit, but I would think this is going to be so bad from a performance standpoint that we might want to throw much sooner. I am open to suggestions, this just feels like it is too small.

Reply:
This is the halting problem.  We don't know ahead of time how expensive this split-and-retry loop will be, and we also don't know how expensive downstream operations will be on increasingly smaller batches.  Thus I don't think any hardcoded limit besides the inherent limit makes sense here if the whole point of these changes is to not fail but succeed despite the performance penalty.

I hate adding yet more configs, but we could add a time-based config for how long we will spend retrying to put a sort of limit on this, but that would not cover the time spent in downstream processing of small batches.

Reply:
This seems similar to the number limit on retry attempts we were looking to add. It feels we need a separate issue that adds a count limit and a time limit (with configs) within a retry block. It seems the retry logic can be adapted for it and it handled in the core framework, or the core framework could choose to end things early by throwing a different exception.